nx/source/services/fspr.c

```diff
@@ -55,141 +38,40 @@ Result fsprRegisterProgram(u64 pid, u64 titleID, FsStorageId storageID, const vo
-    raw->sid = storageID;
-    raw->pid = pid;
-    raw->tid = titleID;
-    raw->fah_size = fah_size;
-    raw->fac_size = fac_size;
-    
-    Result rc = serviceIpcDispatch(&g_fsprSrv);
-    
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
-
-        struct {
-            u64 magic;
-            u64 result;
-        } *resp = r.Raw;
-
-        rc = resp->result;
-    }
-    
-    return rc;
+    } in = { sid, pid, tid, fah_size, fac_size };
+    return serviceDispatchIn(&g_fsprSrv, 0, in,
+        .buffer_attrs = {
+            (SfBufferAttr_In | SfBufferAttr_HipcMapAlias),
+            (SfBufferAttr_In | SfBufferAttr_HipcMapAlias),
+        },
+        .buffers = {
+            { fs_access_header,  fah_size },
+            { fs_access_control, fac_size },
+        },
+    );
 }
 
 Result fsprUnregisterProgram(u64 pid) {
-    IpcCommand c;
-    ipcInitialize(&c);
-    
-    struct {
-        u64 magic;
-        u64 cmd_id;
-        u64 pid;
-    } *raw;
-    
-    raw = ipcPrepareHeader(&c, sizeof(*raw));
-    raw->magic = SFCI_MAGIC;
-    raw->cmd_id = 1;
-    raw->pid = pid;
-    
-    Result rc = serviceIpcDispatch(&g_fsprSrv);
-    
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
-
-        struct {
-            u64 magic;
-            u64 result;
-        } *resp = r.Raw;
-
-        rc = resp->result;
-    }
-    
-    return rc;
+    return serviceDispatchIn(&g_fsprSrv, 1, pid);
 }
 
 Result fsprSetCurrentProcess(void) {
-    IpcCommand c;
-    ipcInitialize(&c);
-    ipcSendPid(&c);
-
-    struct {
-        u64 magic;
-        u64 cmd_id;
-        u64 unk;
-    } *raw;
-
-    raw = ipcPrepareHeader(&c, sizeof(*raw));
-
-    raw->magic = SFCI_MAGIC;
-    raw->cmd_id = 2;
-    raw->unk = 0;
-
-    Result rc = serviceIpcDispatch(&g_fsprSrv);
-
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
-
-        struct {
-            u64 magic;
-            u64 result;
-        } *resp = r.Raw;
-
-        rc = resp->result;
+    if(hosversionBefore(4,0,0)) {
+        return MAKERESULT(Module_Libnx, LibnxError_IncompatSysVer);
     }
-    
-    return rc;
+
+    u64 pid_placeholder = 0;
+    return serviceDispatchIn(&g_fsprSrv, 2, pid_placeholder, .in_send_pid = true);
 }
 
 
 Result fsprSetEnabledProgramVerification(bool enabled) {
-    IpcCommand c;
-    ipcInitialize(&c);
-    
-    struct {
-        u64 magic;
-        u64 cmd_id;
-        u8 enabled;
-    } *raw;
-    
-    raw = ipcPrepareHeader(&c, sizeof(*raw));
-    raw->magic = SFCI_MAGIC;
-    raw->cmd_id = 256;
-    raw->enabled = enabled ? 1 : 0;
-    
-    Result rc = serviceIpcDispatch(&g_fsprSrv);
-    
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
-
-        struct {
-            u64 magic;
-            u64 result;
-        } *resp = r.Raw;
-
-        rc = resp->result;
-    }
-    
-    return rc;
+    return serviceDispatchIn(&g_fsprSrv, 256, enabled);
```

Comment:
Use the input-bool helper funcs from elsewhere, or just impl what those do here.